2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.snowplow.collectors/scalastream/ResponseHandler.scala

```diff
@@ -89,141 +98,144 @@ object ResponseHandler {
-      event.hostname = hostname
-      event.networkUserId = networkUserId
-
-      userAgent.foreach(event.userAgent = _)
-      refererUri.foreach(event.refererUri = _)
-      event.headers = request.headers.flatMap {
-        case _: `Remote-Address` | _: `Raw-Request-URI` => None
-        case other => Some(other.toString)
+    // Make a Tuple2 with the ip address and the shard partition key
+    var (ipAddress, partitionKey) = ip.toOption.map(_.getHostAddress) match {
+      case None => ("unknown", UUID.randomUUID.toString)
+      case Some(ip) => (ip, if (config.useIpAddressAsPartitionKey) ip else UUID.randomUUID.toString)
+    }
+
+    // Check if nuid param is present
+    val networkUserIdParam = request.uri.query.get("nuid")
+    val networkUserId: String = networkUserIdParam match {
+      // Use nuid as networkUserId if present
+      case Some(nuid) => nuid
+      // Else use the same UUID if the request cookie contains `sp`.
+      case None =>  requestCookie match {
+        case Some(rc) => rc.content
+        case None     => UUID.randomUUID.toString
       }
+    }
 
-      // Set the content type
-      request.headers.find(_ match {case `Content-Type`(ct) => true; case _ => false}) foreach {
+    // Construct an event object from the request.
+    val timestamp: Long = System.currentTimeMillis
+
+    val event = new CollectorPayload(
+      "iglu:com.snowplowanalytics.snowplow/CollectorPayload/thrift/1-0-0",
+      ipAddress,
+      timestamp,
+      "UTF-8",
+      Collector
+    )
+
+    event.path = path
+    event.querystring = queryParams
+    event.body = body
+    event.hostname = hostname
+    event.networkUserId = networkUserId
+
+    userAgent.foreach(event.userAgent = _)
+    refererUri.foreach(event.refererUri = _)
+    event.headers = request.headers.flatMap {
+      case _: `Remote-Address` | _: `Raw-Request-URI` => None
+      case other => Some(other.toString)
+    }
 
-        // toLowerCase called because Spray seems to convert "utf" to "UTF"
-        ct => event.contentType = ct.value.toLowerCase
-      }
+    // Set the content type
+    request.headers.find(_ match {case `Content-Type`(ct) => true; case _ => false}) foreach {
 
-      // Only send to Kinesis if we aren't shutting down
-      val sinkResponse = sinks.good.getType match {
-        case Sink.Kinesis if KinesisSink.shuttingDown => null
-        case _ => {
-          // Split events into Good and Bad
-          val eventSplit = SplitBatch.splitAndSerializePayload(event, sinks.good.MaxBytes)
+      // toLowerCase called because Spray seems to convert "utf" to "UTF"
+      ct => event.contentType = ct.value.toLowerCase
+    }
 
-          // Send events to respective sinks
-          val sinkResponseGood = sinks.good.storeRawEvents(eventSplit.good, partitionKey)
-          val sinkResponseBad  = sinks.bad.storeRawEvents(eventSplit.bad, partitionKey)
+    //if kinesisPartKey is specified in header, use it
+    if(kinesisPartKey != null){
```

Comment:
Check to see if the kinesisPartKey exists, if it does, use it. Otherwise, use what it's already been set to (most likely a random guid, see ~lines 112 - 115